CommonTools/PileupAlgos/plugins/PuppiProducer.cc

```diff
@@ -288,9 +294,17 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
       pCand->setSourceCandidatePtr(aCand.sourceCandidatePtr(0));
       fPackedPuppiCandidates->push_back(*pCand);
     } else {
-      pfCand->setP4(puppiP4s.back());
+      pfCandWeighted->setP4(puppiP4s.back());
+      pfCandWeighted->setSourceCandidatePtr(aCand.sourceCandidatePtr(0));
+      pfCand->setP4(aCand.p4());
       pfCand->setSourceCandidatePtr(aCand.sourceCandidatePtr(0));
+      if (fPuppiForLeptons) {
+        pfCandWeighted->setPuppiWeight(1.0, lWeights[iCand]);
+      } else {
+        pfCandWeighted->setPuppiWeight(lWeights[iCand]);
+      }
```

Comment:
I don't quite understand the logic behind this.
My understanding from the code is that the "PFCandidates_puppi__*" collection would contain the original PF candidates and the PUPPI weights, while the "PFCandidates_puppi_weighted_*" collection would contain the scaled candidates and is not used anywhere after the full transition of all downstream code.

From this I would expect here
pfCand->setPuppiWeight
rather than
pfCandWeighted->setPuppiWeight

Did I misunderstand?

Reply:
Shoot, you're right. I will fix it.

Reply:
And also this should keep the "other" weight, like a few lines before in the code, e.g.
pCand->setPuppiWeight(pCand->puppiWeight(), lWeights[iCand]);

Reply:
That would keep the existing puppi weight, which is handled elsewhere, right?